### PR TITLE
Add Network Load Balancer

### PR DESCRIPTION
**Pull request by pierre-emmanuelJ**
Add Network Load Balancer

### CODE REVIEW

**Comment by pierre-emmanuelJ** [comment]
TestAccResourceNlb
```Shell
go test -mod vendor -v -timeout 60m github.com/terraform-providers/terraform-provider-exoscale/exoscale -run '(TestAccResourceNlb)$'
=== RUN   TestAccResourceNlb
--- PASS: TestAccResourceNlb (35.90s)
PASS
ok  	github.com/terraform-providers/terraform-provider-exoscale/exoscale	35.930s
```
TestAccResourceNlbService
```Shell
go test -mod vendor -v -timeout 60m github.com/terraform-providers/terraform-provider-exoscale/exoscale -run '(TestAccResourceNlbService)$'
=== RUN   TestAccResourceNlbService
--- PASS: TestAccResourceNlbService (186.23s)
PASS
ok  	github.com/terraform-providers/terraform-provider-exoscale/exoscale	186.254s
```

File: exoscale/resource_exoscale_nlb_service.go [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+	if d.HasChange("target_port") {
+		service.TargetPort = int64(d.Get("target_port").(int))
+		d.SetPartial("target_port")
+	}
+
+	if d.HasChange("healthcheck") {
+		raw := d.Get("healthcheck").(*schema.Set)
+		healthcheck := raw.List()[0].(map[string]interface{})
+
+		d.SetPartial("healthcheck")
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb_service.go]: No need for partials here, the NLB service is updated atomically.

File: exoscale/provider.go [hunk 2]
```diff
@@ -142,6 +142,8 @@ func Provider() terraform.ResourceProvider {
 			"exoscale_ipaddress":            resourceIPAddress(),
 			"exoscale_network":              resourceNetwork(),
 			"exoscale_nic":                  resourceNIC(),
+			"exoscale_nlb":                  resourceNlb(),
```
Comment (falzm) [inline exoscale/provider.go]: Same as "NIC" and "IP", "NLB" is an acronym therefore it must be capitalized:

* `resourceNLB`
* `resourceNLBService`

(applies everywhere in the code)

File: exoscale/resource_exoscale_nlb.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+	nlb, err := findNlb(ctx, d, meta)
+	if err != nil {
+		return err
+	}
+
+	if nlb == nil {
+		return fmt.Errorf("Network Load Balancer %q not found", d.Id())
+	}
+
+	d.Partial(true)
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb.go]: No need for partials here, the NLB is updated atomically.

File: exoscale/resource_exoscale_nlb.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	}
+	zones := resp.(*egoscale.ListZonesResponse).Zone
+
+	var nlb *egoscale.NetworkLoadBalancer
+	for _, zone := range zones {
+		n, err := client.GetNetworkLoadBalancer(
+			ctx,
+			zone.Name,
+			d.Id(),
+		)
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb.go]: Let's go for the compact format here, since parameters are short it's perfectly readable 😉

```suggestion
		n, err := client.GetNetworkLoadBalancer(ctx, zone.Name, d.Id())
```

**Review by falzm — CHANGES_REQUESTED** [review]

File: exoscale/provider.go (hunk 2 above)
Comment (pierre-emmanuelJ) [reply exoscale/provider.go]: fixed: 77d6edd

File: exoscale/resource_exoscale_nlb.go (hunk 3 above)
Comment (pierre-emmanuelJ) [reply exoscale/resource_exoscale_nlb.go]: 246cbf5

File: exoscale/resource_exoscale_nlb_service.go (hunk 1 above)
Comment (pierre-emmanuelJ) [reply exoscale/resource_exoscale_nlb_service.go]: b2a0010

File: exoscale/resource_exoscale_nlb.go (hunk 4 above)
Comment (pierre-emmanuelJ) [reply exoscale/resource_exoscale_nlb.go]: ccb328f2e6db70d8e23c79b0b2de24f71f472e01

File: examples/nlb/main.tf [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+resource "exoscale_nlb" "nlb-website" {
+  name = "website-load-balancer"
+  description = "This is the Network Load Balancer for my website"
+  zone = var.zone
+}
+
+resource "exoscale_nlb_service" "nlb-service-website" {
+  zone = var.zone
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
  zone = exoscale_nlb.website.zone
```

File: examples/nlb/main.tf [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+resource "exoscale_nlb" "nlb-website" {
+  name = "website-load-balancer"
+  description = "This is the Network Load Balancer for my website"
+  zone = var.zone
+}
+
+resource "exoscale_nlb_service" "nlb-service-website" {
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
resource "exoscale_nlb_service" "website" {
```

File: examples/nlb/main.tf [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  template_id = data.exoscale_compute_template.instancepool.id
+  service_offering = "medium"
+  size = 3
+  disk_size = 50
+  user_data = "#cloud-config\npackage_upgrade: true\n"
+  key_pair = exoscale_ssh_keypair.key.name
+  zone = var.zone
+}
+
+resource "exoscale_nlb" "nlb-website" {
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
resource "exoscale_nlb" "website" {
```

File: examples/nlb/main.tf [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+resource "exoscale_instance_pool" "instancepool-website" {
+  name = "instancepool-website"
+  description = "test"
+  template_id = data.exoscale_compute_template.instancepool.id
+  service_offering = "medium"
+  size = 3
+  disk_size = 50
+  user_data = "#cloud-config\npackage_upgrade: true\n"
+  key_pair = exoscale_ssh_keypair.key.name
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
```

File: examples/nlb/main.tf [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+variable "template" {
+  default = "Linux Ubuntu 18.04 LTS 64-bit"
+}
+
+data "exoscale_compute_template" "instancepool" {
+  zone = var.zone
+  name = var.template
+}
+
+resource "exoscale_instance_pool" "instancepool-website" {
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
resource "exoscale_instance_pool" "website" {
```

File: examples/nlb/main.tf [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+resource "exoscale_ssh_keypair" "key" {
+  name = "mykey-nlb-test"
+}
+
+variable "template" {
+  default = "Linux Ubuntu 18.04 LTS 64-bit"
+}
+
+data "exoscale_compute_template" "instancepool" {
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
data "exoscale_compute_template" "website" {
```

File: examples/nlb/main.tf [hunk 11]
```diff
@@ -0,0 +1,62 @@
+variable "zone" {
+  default = "de-fra-1"
+}
+
+resource "exoscale_ssh_keypair" "key" {
+  name = "mykey-nlb-test"
+}
+
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
```

File: examples/nlb/main.tf [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  service_offering = "medium"
+  size = 3
+  disk_size = 50
+  user_data = "#cloud-config\npackage_upgrade: true\n"
+  key_pair = exoscale_ssh_keypair.key.name
+  zone = var.zone
+}
+
+resource "exoscale_nlb" "nlb-website" {
+  name = "website-load-balancer"
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
  name = "website"
```

File: examples/nlb/main.tf [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  description = "This is the Network Load Balancer for my website"
+  zone = var.zone
+}
+
+resource "exoscale_nlb_service" "nlb-service-website" {
+  zone = var.zone
+  name = "nlb-service-website"
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
  name = "website"
```

File: examples/nlb/main.tf [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    retries = 1
+    uri = ""
+  }
+}
+
+provider "exoscale" {
+  version = "~> 0.15"
+  key = var.key
+  secret = var.secret
+  compute_endpoint = "https://api.exoscale.com/v1"
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
```

File: examples/nlb/main.tf [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    port = 9595
+    interval = "5s"
+    timeout = "5s"
+    retries = 1
+    uri = ""
+  }
+}
+
+provider "exoscale" {
+  version = "~> 0.15"
```
Comment (falzm) [inline examples/nlb/main.tf]:
```suggestion
```

File: exoscale/provider.go [hunk 16]
```diff
@@ -102,6 +102,13 @@ func Provider() terraform.ResourceProvider {
 				Description: fmt.Sprintf("Exoscale DNS API endpoint (by default: %s)", defaultDNSEndpoint),
 				DefaultFunc: schema.EnvDefaultFunc("EXOSCALE_DNS_ENDPOINT", defaultDNSEndpoint),
 			},
+			"environment": {
+				Type:     schema.TypeString,
+				Optional: true,
+				DefaultFunc: schema.MultiEnvDefaultFunc([]string{
+					"EXOSCALE_ENVIRONMENT",
```
Comment (falzm) [inline exoscale/provider.go]: The eventual naming convention for API-related settings is `EXOSCALE_API_*`:

```suggestion
					"EXOSCALE_API_ENVIRONMENT",
```

File: exoscale/provider_test.go [hunk 17]
```diff
@@ -27,11 +27,18 @@ type testAttrs map[string]schema.SchemaValidateFunc
...
 
+var testEnvironment string
+
 func init() {
 	testAccProvider = Provider().(*schema.Provider)
 	testAccProviders = map[string]terraform.ResourceProvider{
 		"exoscale": testAccProvider,
 	}
+
+	testEnvironment = os.Getenv("EXOSCALE_ENVIRONMENT")
```
Comment (falzm) [inline exoscale/provider_test.go]:
```suggestion
	testEnvironment = os.Getenv("EXOSCALE_API_ENVIRONMENT")
```

File: exoscale/client.go [hunk 18]
```diff
@@ -62,3 +64,11 @@ func GetDNSClient(meta interface{}) *egoscale.Client {
 	}
 	return config.dnsClient
 }
+
+func environment(meta interface{}) string {
```
Comment (falzm) [inline exoscale/client.go]: To stay consistent with the other `getXXXX` helpers:

```suggestion
func getEnvironment(meta interface{}) string {
```

File: exoscale/resource_exoscale_nlb.go [hunk 19]
```diff
@@ -0,0 +1,275 @@
...
+	ctx = apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), ""))
+	resp, err := client.RequestWithContext(ctx, egoscale.ListZones{})
+	if err != nil {
+		return nil, err
+	}
+	zones := resp.(*egoscale.ListZonesResponse).Zone
+
+	var nlb *egoscale.NetworkLoadBalancer
+	for _, zone := range zones {
+		n, err := client.GetNetworkLoadBalancer(ctx, zone.Name, d.Id())
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb.go]: Slight change to the implementation: the current zone has to be specified for each iteration:

```suggestion
	resp, err := client.RequestWithContext(ctx, egoscale.ListZones{})
	if err != nil {
		return nil, err
	}
	zones := resp.(*egoscale.ListZonesResponse).Zone

	var nlb *egoscale.NetworkLoadBalancer
	for _, zone := range zones {
		n, err := client.GetNetworkLoadBalancer(
		apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), zone.Name)),
		zone.Name,
		d.Id())
```

File: exoscale/resource_exoscale_nlb_service.go [hunk 20]
```diff
@@ -0,0 +1,417 @@
...
+
+	ctx = apiv2.WithEndpoint(ctx, apiv2.NewReqEndpoint(environment(meta), ""))
+	resp, err := client.RequestWithContext(ctx, egoscale.ListZones{})
+	if err != nil {
+		return nil, err
+	}
+	zones := resp.(*egoscale.ListZonesResponse).Zone
+
+	for _, zone := range zones {
+		nlbs, err := client.ListNetworkLoadBalancers(ctx, zone.Name)
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb_service.go]: Same change described for `findNLB()` has to be done here.

**Review by falzm — CHANGES_REQUESTED** [review]

File: exoscale/resource_exoscale_nlb_service.go (hunk 20 above)
Comment (pierre-emmanuelJ) [reply exoscale/resource_exoscale_nlb_service.go]: https://github.com/terraform-providers/terraform-provider-exoscale/pull/52/commits/e17659e16939890c12f2de6a7ccef9bd13ab3a99

File: exoscale/resource_exoscale_nlb.go [hunk 21]
```diff
@@ -0,0 +1,317 @@
...
+		return err
+	}
+
+	if err := d.Set("ip_address", nlb.IPAddress.String()); err != nil {
+		return err
+	}
+
+	services := make([]string, len(nlb.Services))
+	for i, service := range nlb.Services {
+		services[i] = service.Name
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb.go]: I suggest exposing the service ID instead of the name, so they can be reused in another module or in outputs for additional actions (which require the ID):

```suggestion
		services[i] = service.ID
```

File: exoscale/resource_exoscale_nlb.go [hunk 22]
```diff
@@ -0,0 +1,317 @@
...
+	if nlb == nil {
+		return nil, fmt.Errorf("Network Load Balancer %q not found", d.Id())
+	}
+
+	if err := resourceNLBApply(ctx, client, d, nlb); err != nil {
+		return nil, err
+	}
+
+	resources := make([]*schema.ResourceData, 0, 1+len(nlb.Services))
+	resources = append(resources, d)
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb.go]:
```suggestion
	resources := []*schema.ResourceData{d}
```

File: exoscale/resource_exoscale_nlb_service_test.go [hunk 23]
```diff
@@ -0,0 +1,375 @@
...
+	testAccResourceNLBServiceHealthcheckTimeout     = "5"
+	testAccResourceNLBServiceHealthcheckRetries     = "1"
+	testAccResourceNLBServiceHealthcheckPort        = "8080"
+	testAccResourceNLBServiceHealthcheckPortUpdated = "80"
+	testAccResourceNLBServiceHealthcheckURI         = "/healthz"
+
+	testAccResourceNLBServiceConfigCreate = fmt.Sprintf(`
+resource "exoscale_ssh_keypair" "key" {
+  name = "%s"
+}
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb_service_test.go]: You don't need SSH keys for tests, I suggest you restrict attributes to the strict functional minimum to keep them as short as possible.

File: exoscale/resource_exoscale_nlb_service_test.go [hunk 24]
```diff
@@ -0,0 +1,375 @@
...
+					testAccCheckResourceNLBService(service),
+					testAccCheckResourceNLBServiceAttributes(testAttrs{
+						"name":                            ValidateString(testAccResourceNLBServiceName),
+						"description":                     ValidateString(testAccResourceNLBServiceDescription),
+						"protocol":                        ValidateString(testAccResourceNLBServiceProtocol),
+						"target_port":                     ValidateString(testAccResourceNLBServiceTargetPort),
+						"port":                            ValidateString(testAccResourceNLBServicePort),
+						"strategy":                        ValidateString(testAccResourceNLBServiceStrategy),
+						"healthcheck.#":                   ValidateString("1"),
+						"healthcheck.1241295229.mode":     ValidateString(testAccResourceNLBServiceHealthcheckMode),
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb_service_test.go:203]: What does the `1241295229` value corresponds to? Isn't it possible to use a more generic addressing such as `healthcheck.*.mode`?

File: exoscale/resource_exoscale_nlb_test.go [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+						"description": ValidateString(testAccResourceNLBDescriptionUpdated),
+					}),
+				),
+			},
+			{
+				ResourceName:            "exoscale_nlb.nlb",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"state"},
+				ImportStateCheck: func(s []*terraform.InstanceState) error {
```
Comment (falzm) [inline exoscale/resource_exoscale_nlb_test.go]: You should also check that the services are imported correctly, now that the `exoscale_nlb` resource also imports the child services.

**Review by falzm — CHANGES_REQUESTED** [review]

File: exoscale/resource_exoscale_nlb_service_test.go (hunk 23 above)
Comment (pierre-emmanuelJ) [reply exoscale/resource_exoscale_nlb_service_test.go]: https://github.com/terraform-providers/terraform-provider-exoscale/pull/52/commits/4f89cc839c1a113a3dbcf8dc05cdb24e03775714

File: exoscale/resource_exoscale_nlb_test.go (hunk 25 above)
Comment (pierre-emmanuelJ) [reply exoscale/resource_exoscale_nlb_test.go]: https://github.com/terraform-providers/terraform-provider-exoscale/pull/52/commits/e616277314afa58197f3dc3dd57a20b7424203d3 😄

File: website/docs/r/nlb_service.html.markdown [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+
+## Example Usage
+
+```hcl
+variable "zone" {
+  default = "de-fra-1"
+}
+
+variable "template" {
+  default = "Linux Ubuntu 18.04 LTS 64-bit"
```
Comment (falzm) [inline website/docs/r/nlb_service.html.markdown]:
```suggestion
  default = "Linux Ubuntu 20.04 LTS 64-bit"
```

😏

File: website/docs/r/nlb_service.html.markdown [hunk 27]
```diff
@@ -0,0 +1,96 @@
+---
+layout: "exoscale"
+page_title: "Exoscale: exoscale_nlb_service"
+sidebar_current: "docs-exoscale-nlb-service"
+description: |-
+  Provides an Exoscale Network Load Balancer Service resource.
```
Comment (falzm) [inline website/docs/r/nlb_service.html.markdown]:
```suggestion
  Provides an Exoscale Network Load Balancer service resource.
```

Since it's a noun in this context, *service* should not be capitalized (applies everywhere).

File: website/docs/r/nlb_service.html.markdown [hunk 28]
```diff
@@ -0,0 +1,96 @@
...
+resource "exoscale_nlb" "website" {
+  name = "website"
+  description = "This is the Network Load Balancer for my website"
+  zone = var.zone
+}
+
+resource "exoscale_nlb_service" "website" {
+  zone = exoscale_nlb.website.zone
+  name = "website"
+  description = "This is the Network Load Balancer Service for my website"
```
Comment (falzm) [inline website/docs/r/nlb_service.html.markdown]:
```suggestion
  description = "Website over HTTP"
```

File: website/docs/r/nlb_service.html.markdown [hunk 29]
```diff
@@ -0,0 +1,96 @@
...
+	port = 80
+	target_port = 8080
+	strategy = "round-robin"
+
+  healthcheck {
+    port = 8080
+    mode = "http"
+    uri = "/healthz"
+    interval = 5
+    timeout = 5
```
Comment (falzm) [inline website/docs/r/nlb_service.html.markdown]: Timeout must be lower than interval.

```suggestion
    timeout = 3
```

File: website/docs/r/nlb_service.html.markdown [hunk 30]
```diff
@@ -0,0 +1,96 @@
...
+    retries = 1
+  }
+}
+```
+
+## Argument Reference
+
+* `nlb_id` - (Required) The ID of the NLB to attach the Service.
+* `zone` - (Required) The name of the [zone][zone] used by the NLB.
+* `instance_pool_id` - (Required) The ID of the Instance Pool to attach the service.
```
Comment (falzm) [inline website/docs/r/nlb_service.html.markdown]:
```suggestion
* `instance_pool_id` - (Required) The ID of the Instance Pool to forward network traffic to.
```

File: website/docs/r/nlb_service.html.markdown [hunk 31]
```diff
@@ -0,0 +1,96 @@
...
+* `name` - (Required) The name of the NLB Service.
+* `port` - (Required) The port of the NLB service.
+* `target_port` - (Required) The port targeted by the NLB to the Instance Pool.
```
Comment (falzm) [inline website/docs/r/nlb_service.html.markdown]:
```suggestion
* `target_port` - (Required) The port to forward network traffic to on target instances.
```

File: website/docs/r/nlb_service.html.markdown (hunk 27 above)
Comment (pierre-emmanuelJ) [reply website/docs/r/nlb_service.html.markdown]: 4bd3625

**Review by falzm — APPROVED** [review]
LGTM. Let's hold on for a little while before merging, since the egoscale package will have to be updated following recent API changes.
